Add controls to start and stop camera recording

haotongwang opened the pull request:
## Description

Add Start and Stop recording buttons to publish start (`/v3/camera/recording/start)` and stop (`/v3/camera/recording/stop`) topics for starting and stopping a camera's recording. Also includes status displays for both cameras.

Initally displays the last received payload from `/v3/camera/recording/status/<primary/secondary>` if it exists.

## Screenshots

![off](https://user-images.githubusercontent.com/50545432/88550872-87646780-d065-11ea-9beb-02290b2fc3ab.png)

![recording](https://user-images.githubusercontent.com/50545432/88550936-99460a80-d065-11ea-912b-33993092ed2b.png)

![error](https://user-images.githubusercontent.com/50545432/88551055-be3a7d80-d065-11ea-9249-f28f697cd9cd.png)

## Steps to Test

Start `mosquitto` broker

Start [raspicam](https://github.com/monash-human-power/raspicam) with `python overlay_new.py` (or any over really)

Start [server](../server) and [client](../client)

### Tests

- Wait for raspicam to automatically publish
- Run `mosquitto_pub -t /v3/camera/recording/status/primary -m "{\"status\":\"recording\", \"recordingMinutes\": 10.50, \"recordingFile\": \"test.mp4\", \"diskSpaceRemaining\": 100000000000}"`

haotongwang committed:
Add CameraRecording component skeleton

Include component in CameraSettingsView

haotongwang committed:
Implement mqtt for dashboard

Add socket.io emit from client for starting and stopping recording
Add socket.io listen for updating status
Add mqtt publish from server for start and stop
Subscribe server to /v3/camera/recording/status/<primary, secondary>

Bug:

- Server socket.emit (or client CameraRecording useChannel) does not
seem to send (or receive) messages

haotongwang committed:
Add status messages to dashboard

Implement different information for each status
Add util function for representing bytes in a more readable unit

haotongwang committed:
Generalise CameraRecording with CameraRecordingStatus

CameraRecording is a wrapper for multiple CameraRecordingStatus
CameraRecordingStatus is the status for a single camera
Generalise mqtt code for server for catching camera recording statuses

haotongwang committed:
Implement session recording status

Display the last received camera recording payload if it exists

Wrap CameraRecordingStatus in div instead of p

khanguslee commented:
wow. Good job! How the heck did u make a PR this fast??? 💯

haotongwang commented:
Haha spent most of today on it and kept spamming Chris with questions.
~~I accidentally did "comment and close", don't know how to reopen it :/~~ I think I fixed it

emilytrau commented on client/src/components/v2/CameraRecording.js:
```diff
@@ -0,0 +1,56 @@
+ * @returns {React.Component<CameraRecordingProps>} Component
+ */
+export default function CameraRecording({ devices }) {
+  const startRecording = () => {
```
These helper functions should be in a module in the api folder. The reason for this is that this allows us to provide a stable interface to the display components by abstracting away the implementation details of the API. For instance, V3 may replace "emit" with another method of communication.

haotongwang replied:
Done
480c632

haotongwang commented:
test

emilytrau commented on client/src/components/v2/CameraRecordingStatus.js:
```diff
@@ -0,0 +1,93 @@
+   * @param {string} payload Payload structure is defined in https://www.notion.so/V3-MQTT-Topics-66e6715d0e1941ffaa82020e5868fbae. See /v3/camera/recording/status/<primary/>secondary>
+   * @returns {React.Component} Component
+   */
+  function formatPayload(payload) {
```
Parsing the payload should be done within the API library for a similar reason.

haotongwang replied:
Done
1442708

`formatPayload` has been split (see #74 (comment)) and moved to the API library

emilytrau commented on client/src/components/v2/CameraRecordingStatus.js:
```diff
@@ -0,0 +1,93 @@
+  }
+
+  const lastPayload = sessionStorage.getItem(`camera-recording-last-payload-${device}`);
+  const [status, setStatus] = useState((
```
Status should just be the status string, rather than a component. This allows for easier debugging using the React dev tools. Any component creation should strictly happen within the render only.

haotongwang replied:
Done
1442708

emilytrau commented on client/src/components/v2/CameraRecordingStatus.js:
```diff
@@ -0,0 +1,93 @@
+    setStatus(formatPayload(payload));
+  }, [device]);
+
+  useChannel(`camera-recording-status-${device}`, updateStatus);
```
This can be abstracted into a custom hook within the API library.

haotongwang replied:
Done
2a1126d

Custom react hooks are pretty cool!

emilytrau commented on client/src/components/v2/CameraRecording.js:
```diff
@@ -0,0 +1,56 @@
+      <Card.Body>
+        <Card.Title>Recording Controls</Card.Title>
+        {
+          devices.split(',').map((device) => (
```
You can pass in an array of strings for devices here to make the format of the argument clearer and reduce the overhead of splitting.

haotongwang replied:
Done
https://github.com/monash-human-power/dashboard/pull/74/commits/7c9a27d57887c765470f6eaf8879817115defbc9

I should have known that React would have a way to pass objects as props instead of just a string haha.

emilytrau commented on client/src/components/v2/CameraRecordingStatus.js:
```diff
@@ -0,0 +1,93 @@
+import React, { useCallback, useState } from 'react';
+import { useChannel, emit } from 'api/v2/socket';
+import formatBytes from 'utils/formatBytes';
+// import { Card } from 'react-bootstrap';
```
Remove unused imports

haotongwang replied:
Done
https://github.com/monash-human-power/dashboard/pull/74/commits/6587f234f8683fb768efeef951209b5a7e8d8082

emilytrau reviewed (changes requested):
Great job on your very first PR! This is already looking really good!

khanguslee commented on client/src/components/v2/CameraRecording.js:
```diff
@@ -0,0 +1,56 @@
+        <Card.Title>Recording Controls</Card.Title>
+        {
+          devices.split(',').map((device) => (
+            <div style={{ marginBottom: '10px' }} key={device}>
```
Inline stylings. Angus says no 🙅 . Please put your styles inside a `.module.css` file.

haotongwang replied:
Done
d6ae855
54673ff

khanguslee commented on client/src/views/v2/CameraSettingsView.js:
```diff
@@ -10,6 +11,9 @@ import CameraSettings from 'components/v2/CameraSettings';
   return (
     <ContentPage title="Camera Settings">
+      <div className="mb-4">
+        <CameraRecording devices="primary,secondary" />
```
You wouldn't have that nasty string split inside `CameraRecording` if you had just passed this in as an array.

`        <CameraRecording devices=["primary", "secondary"] />`

haotongwang replied:
Done
https://github.com/monash-human-power/dashboard/pull/74/commits/7c9a27d57887c765470f6eaf8879817115defbc9

I should have known that React would have a way to pass objects as props instead of just a string haha.

khanguslee commented on client/src/components/v2/CameraRecording.js:
```diff
@@ -0,0 +1,56 @@
+      <Card.Body>
+        <Card.Title>Recording Controls</Card.Title>
+        {
+          devices.split(',').map((device) => (
```
See comment in `CameraSettingsView` on how we can get rid of this `.split`

haotongwang replied:
Done
https://github.com/monash-human-power/dashboard/pull/74/commits/7c9a27d57887c765470f6eaf8879817115defbc9

I should have known that React would have a way to pass objects as props instead of just a string haha.

khanguslee commented on client/src/components/v2/CameraRecording.js:
```diff
@@ -0,0 +1,56 @@
+        {
+          devices.split(',').map((device) => (
+            <div style={{ marginBottom: '10px' }} key={device}>
+              <Card.Subtitle>{device[0].toUpperCase() + device.substring(1)}</Card.Subtitle>
```
As a dev reading this, I don't know what `device[0].toUpperCase() + device.substring(1)` represents so it would be good if this was put into a variable before we put it here.

haotongwang replied:
Done
6d68ef1

khanguslee commented on client/src/components/v2/CameraRecording.js:
```diff
@@ -0,0 +1,56 @@
+ *
+ * Starts/stops recording for both displays.
+ *
+ * This is a feature intended for V3 but is current in V2 for testing.
```
nit: `current` -> `currently`

haotongwang replied:
Done
ee481c3

khanguslee commented on client/src/components/v2/CameraRecordingStatus.js:
```diff
@@ -0,0 +1,93 @@
+  /**
+   * Format a camera recording status payload into JSX
+   *
+   * @param {string} payload Payload structure is defined in https://www.notion.so/V3-MQTT-Topics-66e6715d0e1941ffaa82020e5868fbae. See /v3/camera/recording/status/<primary/>secondary>
```
I think we should remove `notion` links because this repo is public LOL. Sorry about this.

minh-tn replied:
You can replace it with " 'V3 MQTT Topics' page on Notion". People will know to use the search function to find the page.

haotongwang replied:
Done
6f3143a

khanguslee commented on client/src/components/v2/CameraRecordingStatus.js:
```diff
@@ -0,0 +1,93 @@
+        info = <p>{data.error}</p>;
+        break;
+
+      default:
```
Good work on adding a default scenario for a switch case. I don't think a lot of people remember to do this..

khanguslee commented on client/src/components/v2/CameraRecordingStatus.js:
```diff
@@ -0,0 +1,93 @@
+  }
+
+  const lastPayload = sessionStorage.getItem(`camera-recording-last-payload-${device}`);
+  const [status, setStatus] = useState((
```
Instead of storing html elements into `status`, why don't we actually store `lastPayload` instead? Ideally, the return function of this functional component should contain _most_ of the view. The logic that is displayed on this view, should be calculated in functions or stored in state beforehand.

This allows future devs to easily see what the main structure of this component looks like, just by looking at the return value.

haotongwang replied:
Done
1442708

Having the return be most of the view does make everything a lot more understandable!

khanguslee commented on client/src/components/v2/CameraRecordingStatus.js:
```diff
@@ -0,0 +1,93 @@
+   * @param {string} payload Payload structure is defined in https://www.notion.so/V3-MQTT-Topics-66e6715d0e1941ffaa82020e5868fbae. See /v3/camera/recording/status/<primary/>secondary>
+   * @returns {React.Component} Component
+   */
+  function formatPayload(payload) {
```
This whole section is a bit confusing to me because it feels like it's trying to do multiple things at once and I just get confused? See below on how I might improve this.

haotongwang replied:
Done
1442708
See https://github.com/monash-human-power/dashboard/pull/74#discussion_r461242513

khanguslee commented on client/src/components/v2/CameraRecordingStatus.js:
```diff
@@ -0,0 +1,93 @@
+  emit('camera-recording-init');
+
+  return (
+    <div>{status}</div>
```
```js
return(
<div>
  <p>Status: {data.status}</p>
  <div>
    <p>{info}</p>
  </div>  
</div>
)
```

`info` can contain the disk space, errors etc.

Using this format, I can clearly see that this component has a status at the top and there's info as the body of the component. If i want to find out how I get this `info`, I look around this component to see where I obtain this info etc.

haotongwang replied:
Done
1442708

I've split `formatPayload` into `getStatusFromPayload` (because `payload` is a string and `formatPayload` is what originally parsed it into JSON) and `getInfoFromPayload`.

khanguslee commented on server/js/sockets.js:
```diff
@@ -91,6 +91,7 @@ sockets.init = function socketInit(server) {
   mqttClient.subscribe('power_model/recommended_SP');
   mqttClient.subscribe('power_model/plan_generated');
   mqttClient.subscribe('camera/push_overlays');
+  // camera recording status subscription occurs when 'camera-recording-init' is received
```
My own nit: Comments starting with capital letters. Not sure why it wasn't picked up by eslint? https://eslint.org/docs/rules/capitalized-comments

haotongwang replied:
Done
130e0c0

Not sure why it wasn't picked up either

khanguslee commented on server/js/sockets.js:
```diff
@@ -149,6 +150,14 @@ sockets.init = function socketInit(server) {
+      } else if (
+        topicString.slice(0, -1).join('/') === '/v3/camera/recording/status'
+      ) {
+        // send mqtt payload on corresponding channel
```
Same here.

haotongwang replied:
Done
130e0c0

Not sure why it wasn't picked up either

khanguslee commented on client/src/utils/formatBytes.js:
```diff
@@ -0,0 +1,14 @@
+  const units = ['Bytes', 'KiB', 'MiB', 'GiB', 'TiB', 'PiB', 'EiB', 'ZiB', 'YiB'];
+  const d = digits < 0 ? 0 : digits;
+  const mag = Math.floor(Math.log(bytes) / Math.log(1024));
+  return `${parseFloat((bytes / (1024 ** mag)).toFixed(d))} ${units[mag]}`;
```
dis some @hallgchris python magick right here

haotongwang replied:
It was shamelessly taken from StackOverflow
cf7c298

khanguslee reviewed (changes requested):
Congrats on your first PR! So fast. Probably faster than @hallgchris first PR too!

Next time, I think it might be better if you had explained what to run in `raspicam`? In this case, the user just needed to run any overlay file.

Anyway, LMK if you have any questions/disagree with my comments! Definitely impressed you didn't know React and already have made your first PR!

khanguslee reviewed (changes requested):
Congrats on your first PR! So fast. Probably faster than @hallgchris first PR too!

Next time, I think it might be better if you had explained what to run in `raspicam`? In this case, the user just needed to run any overlay file.

Anyway, LMK if you have any questions/disagree with my comments! Definitely impressed you didn't know React and already have made your first PR!

haotongwang committed:
Remove unused imports in CameraRecordingStatus.js

haotongwang committed:
Refactor device prop of CameraRecording to be an array instead of a string

haotongwang committed:
Use capitalize from lodash instead of explicit capitalisation

haotongwang committed:
Fix typo in CameraRecording

haotongwang committed:
Remove notion link in CameraRecordingStatus for formatPayload

haotongwang committed:
Add acknowledgement to formatBytes

haotongwang committed:
Capitalise comments in socket.js

haotongwang committed:
Move CameraRecording helper functions into api

haotongwang committed:
Refactor inline styling in CameraRecording into CameraRecording.module.css

haotongwang committed:
Fix typo in import

haotongwang committed:
Refactor CameraRecordingStatus

- Split formatPayload into getStatusFromPayload and getInfoFromPayload
- lastPayload uses useState and storeLastPayload
- JSX in CameraRecordingStatus and getInfoFromPayload is more structurally clear
- getInfoFromPayload dynamically fills in information for fields

haotongwang committed:
Move functionality of CameraRecordingStatus to API

Use custom react hook for status and lastPayload
Abstract the message to server to init camera recording status to API

haotongwang reviewed (commented):
_Okay, I didn't realise this was how you do a review so a lot of my responses are missing from this. Oh well._

Thanks to @angustrau and @khanguslee for the reviews and suggestions. My late-night code looks a lot better now! Please do give more suggestions if you see anything :).

emilytrau commented on client/src/components/v2/CameraRecording.js:
```diff
@@ -0,0 +1,50 @@
+      </Card.Body>
+      <Card.Footer>
+        <Button variant="outline-success" onClick={startRecording}>Start</Button>
+        {' '}
```
You should use a margin here instead of a blank space.

haotongwang replied:
Done
3d22217

emilytrau commented on client/src/components/v2/CameraRecordingStatus.js:
```diff
@@ -0,0 +1,37 @@
+  return (
+    <div>
+      <p>
+        Status:
```
Use a template string to combine these strings explicitly

haotongwang replied:
Done
dd63cb0

emilytrau commented on client/src/api/v2/camera.js:
```diff
@@ -39,3 +41,140 @@ export function useOverlays(device) {
+/**
+ * Initiate receiving camera recording statuses
+ */
+export function initCameraStatus() {
```
This can be called within the useCameraRecordingStatus hook. (Although be wary that hooks are called on every render and this may cause emit to be triggered more than expected)

haotongwang replied:
`initCameraStatus` is now used in a `useEffect` callback with no dependencies in `useCameraRecordingStatus` as you recommended.
4376700

emilytrau commented on client/src/api/v2/camera.js:
```diff
@@ -39,3 +41,140 @@ export function useOverlays(device) {
+ */
+export function useCameraRecordingStatus(device) {
+  const [lastPayload, setLastPayload] = useState(getLastPayload(device));
+  const [status, setStatus] = useState(getStatusFromPayload(lastPayload) || 'Waiting for status...');
```
Since status is always computed from the state of lastPayload it does not need to be a state.

haotongwang replied:
Done
f5a68e5

hallgchris commented on server/js/sockets.js:
```diff
@@ -91,6 +91,7 @@ sockets.init = function socketInit(server) {
   mqttClient.subscribe('power_model/recommended_SP');
   mqttClient.subscribe('power_model/plan_generated');
   mqttClient.subscribe('camera/push_overlays');
+  // Camera recording status subscription occurs when 'camera-recording-init' is received
```
What's the reason for not subscribing by default?

haotongwang replied:
If it subscribes by default, the MQTT response containing the retained payload can (and is often) sent without the client being able to receive it (e.g. still loading). The workaround I decided to go with was to subscribe when the camera page is opened.

haotongwang replied:
Update: the workaround is now subscribing on `io` connection which is at the same time as when the `mqttClient` message handler is set.
4376700

hallgchris commented on client/src/components/v2/CameraRecordingStatus.js:
```diff
@@ -0,0 +1,37 @@
+export default function CameraRecordingStatus({ device }) {
+  const { status, lastPayload } = useCameraRecordingStatus(device);
+
+  initCameraStatus();
```
This cause us to subscribe to the MQTT topics every time a component is loaded. Is there a reason we can't just subscribe when the server loads initially, like for other topics?

emilytrau replied:
This is a socket.io subscription, it's lightweight enough that it doesn't really matter that we sub/unsub. This is an architectural issue due to how I set up the API under the hood, and I'm looking to fix this in V3.

emilytrau replied:
@haotongwang this is a great example of something we can change thanks to a solid abstraction

haotongwang replied:
@angustrau could you elaborate a bit more on what a socket.io subscription is? I added the `initCameraStatus` mainly to not be dependent on `emit` (as you mentioned in an earlier comment) and currently, it just sends a message to the server which subscribes to the topics.

hallgchris replied:
@angustrau When `initCameraStatus` is called, all it does is trigger the server to subscribe to some MQTT topics (right now line 225 of `sockets.js`). The server shouldn't need to subscribe every time the component is loaded.

haotongwang replied:
@hallgchris @angustrau The server now subscribes once on io connection which is at the same time as when the mqttClient message handler is set. If it is subscribed with the other topics, the retained payload would have to be handled by another message handler as it arrives before the main message handler can be set.
4376700

hallgchris commented on client/src/components/v2/CameraRecordingStatus.js:
```diff
@@ -0,0 +1,37 @@
+      </p>
+      <div>
+        <p>{getInfoFromPayload(lastPayload)}</p>
+      </div>
```
Noob question, why does this `p` tag need it's own `div`?

haotongwang replied:
Ah that was left over from before when I copied Angus Lee's JSX. It doesn't change the look of the html so I didn't pick it up.
005d00b

hallgchris commented on client/src/api/v2/camera.js:
```diff
@@ -39,3 +41,140 @@ export function useOverlays(device) {
+  const update = useCallback((newPayload) => {
+    // update last payload
+    storeLastPayload(device, newPayload);
+    setLastPayload(newPayload);
```
Having trouble getting my head around this, what's the difference between `storeLastPayload` and `setLastPayload` here?

hallgchris replied:
Also, should we be storing the most recent statuses from the cameras serverside? Because the most recent status in the `sessionStorage` isn't necessarily the most up to date, there could have been more statuses while the website was not open.

haotongwang replied:
> Having trouble getting my head around this, what's the difference between `storeLastPayload` and `setLastPayload` here?

`storeLastPayload` has been removed as it is now handled server-side. `setLastPayload` is used to update the state of `lastPayload`.

4376700

hallgchris reviewed (changes requested):
Awesome stuff getting this done so quick!

I wanted to test this using an actual Pi to record, but ending up spending a while getting raspicam set up on it. Will update when that's done.

> I didn't realise this was how you do a review so a lot of my responses are missing

No worries, actually we usually respond to PRs in the way you were doing before, to each individual comment, unless you're saying something more general, so no harm done!

> Probably faster than @hallgchris first PR too!

Wow didn't know this was a competition

emilytrau commented on client/src/components/v2/CameraRecordingStatus.js:
```diff
@@ -0,0 +1,37 @@
+ * @returns {React.Component<CameraRecordingStatusProps>} Component
+ */
+export default function CameraRecordingStatus({ device }) {
+  const { status, lastPayload } = useCameraRecordingStatus(device);
```
Since payload can be considered a protocol-specific unstable API, it's best not to depend on it from the display components. It's only used once here, so this can be refactored to expose the payload info rather than the payload itself

haotongwang replied:
I'm not sure what the problem is here. If there is no payload, then `getInfo/StatusFromPayload` will simply return `null`. Could you rephrase or elaborate upon the issue?

emilytrau replied:
What I mean is that anything outside of the API shouldn't rely on the shape of payload. It shouldn't have access to payload at all, since the actual contents of the payload is subject to change. Instead, the display components only have access to processed output of the API, as that can remain stable. When you use `useCameraRecordingStatus` in the display component, you don't actually care what the payload is, just the status, so why return the payload? Everything regarding the payload can be wrapped up inside the API.

haotongwang replied:
`useCameraRecordingStatus` returns type `CameraRecordingStatusPayload` which is the payload string parsed and formatted into an object.
4376700

emilytrau commented on client/src/api/v2/camera.js:
```diff
@@ -39,3 +41,140 @@ export function useOverlays(device) {
+ * @param {string} device Camera
+ * @returns {string} Payload
+ */
+export function getLastPayload(device) {
```
Not every function needs to be exported. In fact it's better to export as few as possible, as any exported functions become part of the public API and might require more refractors down the line if any breaking changes come along. I'd encourage you to only export any methods required by the display components, and keep any other helper methods within this API.

haotongwang replied:
Removed helper functions from public API interface.
4376700

emilytrau commented on client/src/components/v2/CameraRecording.module.css:
```diff
@@ -0,0 +1,3 @@
+.device {
+    margin-bottom: '10px'
```
Semicolon required after this line. Not sure why eslint didn't catch it

haotongwang replied:
Done
3d222178399214dae18154fccb613f141d92209d

Not sure why eslint isn't catching it either

emilytrau commented on client/src/components/v2/CameraRecording.js:
```diff
@@ -0,0 +1,50 @@
+        {
+          devices.map((device) => (
+            <div styles={styles.device} key={device}>
+              <Card.Subtitle>{capitalize(device)}</Card.Subtitle>
```
I'm wary of importing from lodash and other external libraries where possible to reduce bloat and bundle size. This isn't the only instance where we want to derive a pretty name from the camera display name. See [this](https://github.com/monash-human-power/dashboard/blob/e92c459fe32fb620c96bcd2d53a90f7f971b0ae0/client/src/components/v2/CameraSettings.js#L21). Now that it's more useful it would be handy to refactor this in to a "getPrettyDeviceName" API function.

haotongwang replied:
Done
da1d043

haotongwang replied:
Actually done now
4920ea2

emilytrau commented on client/src/api/v2/camera.js:
```diff
@@ -39,3 +41,140 @@ export function useOverlays(device) {
+    return `${name}: ${value}`;
+  };
+
+  return (
```
Methods in the API just transform data. The rendering part of this function should instead be a separate display component.

haotongwang replied:
Done
8a8d1bc

emilytrau reviewed (changes requested):
You're quick! Here's a few more things...

haotongwang committed:
Remove redundent elements in CameraRecordingStatus; add keys to getInfoFromPayload

haotongwang committed:
Refactor styling in CameraRecording

Remove redundent styling
Change blank space to css

haotongwang committed:
Use template string in CameraRecordingStatus

haotongwang committed:
Remove return status in useCameraRecordingStatus

haotongwang committed:
Remove lodash dependency

haotongwang committed:
Remove JSX in API

getInfo/StatusFromPayload -> formatPayload
rendering of data is done in CameraRecordingStatus

hallgchris commented:
Can confirm it is working fine with the RPi!

![image](https://user-images.githubusercontent.com/17876556/88739721-8ffc9100-d17e-11ea-8324-adf9c2f251f1.png)

A few comments having seen this output:
- Seconds obviously need formatting
- Spacing seems a bit unintuitive: Perhaps have a larger space underneath "Primary" and smaller spacing beneath each bit of data shown?
- Would it be feasible to have the primary/secondary stuff displayed as two columns? With both overlays recording, that card would get quite long.

haotongwang committed:
Move storing last payload to server

haotongwang committed:
Extract device name capitalisation to an api function

haotongwang committed:
Update display format

haotongwang committed:
Add columns to display

haotongwang commented:
Updated the display as described by @hallgchris a4d3c05

![Columns](https://user-images.githubusercontent.com/50545432/88908166-e5c75b00-d29c-11ea-8dcd-5872415e64dd.png)

![Rows](https://user-images.githubusercontent.com/50545432/88908222-f081f000-d29c-11ea-9ef3-bc822dfd655b.png)

_This took me wayyyy too long_

I can't seem to replicate the seconds formatting (_works on my end_ <sup>TM</sup>).

nix-chu commented:
@haotongwang what's the last screenshot for at the bottom of your most recent comment?

I also like how Primary and Secondary are now easier to see 👍

hallgchris commented on client/src/api/v2/camera.js:
```diff
@@ -109,7 +97,7 @@ export function formatPayload(payload) {
           {
             const mins = Math.floor(data.recordingMinutes);
             const secs = (data.recordingMinutes - mins) * 60;
```
What data are you testing it with? The pi outputs some nasty numbers:
```json
{"status": "recording", "recordingMinutes": 1.0021132985750834, "recordingFile": "/home/pi/raspicam/recordings/rec_5.h264", "diskSpaceRemaining": 10909519872}
```
Admittedly, those sigfigs are definitely unnecessary and could probably be truncated a bit on the Pi's end.

haotongwang replied:
Ah, I used 2 sig. figs to test and didn't think about the data the Pi would send.
Fixed by flooring seconds value 0e83f6a.

Display for `"recordingMinutes": 1.0021132985750834`

![image](https://user-images.githubusercontent.com/50545432/89000476-d4349080-d33a-11ea-95a6-a8b6e4732d56.png)

hallgchris commented on server/js/sockets.js:
```diff
@@ -11,6 +11,9 @@ const os = require('os');
 let PUBLIC_MQTT_CLIENT;
 
+// global vars
+global.lastRecordingPayloads = {};
```
Just highlighting this for other reviewers, are global vars okay or is there a more correct way?

haotongwang replied:
@angustrau Suggested to use global vars for now and then it can be refactored when client gets refactored to use mqtt directly

hallgchris replied:
Sounds good!

emilytrau replied:
Does it need to be global or is having it scoped just to sockets.js ok?

hallgchris reviewed (changes requested):
Looking fantastic now, that made a big difference!

Looks good for me, only thing is the seconds formatting. I think you might be giving it data that's too neat.

@nix-chu The second photo is showing that when the screen is narrow, the card collapses into a single column (so it'll work nicely on mobile)

haotongwang commented:
> @haotongwang what's the last screenshot for at the bottom of your most recent comment?
> 
> I also like how Primary and Secondary are now easier to see 👍

@nix-chu To show what the display looks like on mobile (when the columns are wrapped to rows)

haotongwang committed:
Fix seconds sig figs

hallgchris reviewed (dismissed):
Looks great, congrats on your first PR! :tada:

emilytrau commented on client/src/api/v2/camera.js:
```diff
@@ -39,3 +41,117 @@ export function useOverlays(device) {
+  if (!data) return null;
+
+  return Object.keys(data)
+    .reduce((acc, field) => {
```
I'm not a fan of reduce... In my opinion a for-loop would give more clarity.

haotongwang replied:
Refactored `reduce` to `forEach`
9c115e8

emilytrau commented on client/src/api/v2/camera.js:
```diff
@@ -39,3 +41,117 @@ export function useOverlays(device) {
+  // only run init once per render
+  useEffect(() => {
+    initCameraStatus();
+    return () => {};
```
The leave callback is optional. You don't need to return anything if you don't use it.

haotongwang replied:
Done
4fcddfd

emilytrau commented on client/src/api/v2/camera.js:
```diff
@@ -39,3 +41,117 @@ export function useOverlays(device) {
+ * Hardcoded for efficiency. If you don't like it, complain to Angus Trau :).
+ *
+ * @param {'primary' | 'secondary'} device Device
+ * @returns {'Primary' | 'Secondary'} Prettied device name
```
In terms of semantics, this function would return a string rather than a specific value type. Anything that uses this function would expect a string type, not anything narrower. This is in contrast to the parameter, where the caller is expected to give 1 of 2 values, and a string type would be too broad.

haotongwang replied:
Done
31facd8

emilytrau commented on client/src/components/v2/CameraRecording.js:
```diff
@@ -0,0 +1,55 @@
+
+/**
+ * @typedef {object} CameraRecordingProps
+ * @property {Array} devices Array of strings of device names
```
This type can be explicitly declared as an array of strings. https://stackoverflow.com/questions/14611995/how-to-specify-an-array-of-objects-as-a-parameter-or-return-value-in-jsdoc

haotongwang replied:
Done
31facd8

I assume we're using JSDoc 3

emilytrau commented on client/src/components/v2/CameraRecording.js:
```diff
@@ -0,0 +1,55 @@
+              <Col
+                className={styles.topMargin}
+                key={device}
+                sm={Math.floor(12 / devices.length)}
```
This can simply be a value of 6. In the case of a new device, it will simply overflow in to a new line. If you want all the devices to fit on exactly 1 line, this can be implemented with a CSS flexbox.

haotongwang replied:
Done
924f560

emilytrau commented on client/src/components/v2/CameraRecording.module.css:
```diff
@@ -0,0 +1,17 @@
+.footer {
+  display: flex;
+  align-items: center;
+  justify-content: center
```
Missing semicolon

haotongwang replied:
Done
4a4ef62

emilytrau commented on client/src/components/v2/CameraRecordingStatus.js:
```diff
@@ -0,0 +1,44 @@
+          ? Object.keys(payload)
+            .map((field) => (
+              <div
+                className={`${styles.bottomMargin} ${styles.flex}`}
```
The `classnames` module is installed and used just for this purpose (to combine multiple classes).

haotongwang replied:
Combined `bottomMargin` and `flex` styles to `statusRow`
869d24f

emilytrau commented on client/src/components/v2/CameraRecordingStatus.module.css:
```diff
@@ -0,0 +1,11 @@
+.bottomMargin {
```
Class names that relate to the component, rather than the effect applied, are preferred. It's a bit more descriptive and obvious when looking at the CSS and figuring out what-does-what

haotongwang replied:
Done
869d24f

haotongwang committed:
Refactor reduce to forEach

haotongwang committed:
Remove redundent callback in useCameraRecordingStatus

haotongwang committed:
Fix JSDoc

haotongwang committed:
Change CameraRecording Col sm to 6 rather than adaptive

haotongwang committed:
;

haotongwang committed:
Use classnames module in CameraRecordingStatus

haotongwang committed:
Rename css class name to relate to component

khanguslee commented on client/src/components/v2/CameraRecording.js:
```diff
@@ -0,0 +1,55 @@
+ */
+export default function CameraRecording({ devices }) {
+  return (
+    <Card>
```
This is SOOO much easier to read now. Good work!

khanguslee commented on client/src/utils/string.js:
```diff
@@ -0,0 +1,20 @@
+ * @returns {string} Start case string
+ */
+export function camelCaseToStartCase(string) {
+  const split = string.replace(/([A-Z])/g, ' $1');
```
Would be good to have unit tests created for these util functions in another PR.

haotongwang replied:
Added as a board card/entry/whatever-it's-meant-to-be-called
https://www.notion.so/Write-unit-tests-for-DAShboard-client-util-functions-870703bfb08d4dea800511e387e13b8a

khanguslee reviewed (dismissed):
LGTM! Congrats!

hallgchris reviewed (dismissed):
Looks good from me! We shall talk about what's next :)

angustrau commented on client/src/components/v2/CameraRecording.js:
```diff
@@ -0,0 +1,55 @@
+          }
+        </Row>
+      </Card.Body>
+      <Card.Footer className={styles.footer}>
```
These classes aren't actually applied...

emilytrau commented on client/src/components/v2/CameraRecording.module.css:
```diff
@@ -0,0 +1,17 @@
+.footer {
+  display: flex;
+  align-items: center;
+  justify-content: center;
```
Bootstrap buttons aren't usually centered. Personally I think it looks better aligned left

haotongwang replied:
![image](https://user-images.githubusercontent.com/50545432/89244935-13b8f080-d64b-11ea-8e30-57bb0a1657f8.png)
f49208f

emilytrau commented on client/src/api/v2/camera.js:
```diff
@@ -39,3 +41,119 @@ export function useOverlays(device) {
+  const data = JSON.parse(payload);
+  if (!data) return null;
+
+  const fData = {};
```
This variable name could be more descriptive

haotongwang replied:
`fData` -> `formattedData`
576b7cf

emilytrau reviewed (changes requested):
This is so close!

haotongwang committed:
Move recording controls buttons left

haotongwang committed:
More descriptive variable names

emilytrau reviewed (approved):
LGTM! Excellent work :+1:

haotongwang commented:
After 37 commits and 120 comments, I can finally merge my first MHP PR!
Thank you to @angustrau @hallgchris and @khanguslee for the reviews!
🎉🎉🎉
